1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+(In fact, there can certainly be scenarios where this is impossible in principle, like remoting and serialization, but there may be tradeoffs and workarounds)
+
+### Context / Scope
+
+The `Engine` should take into account Context / Scope.
+
+Since variables can be copied to other scopes, the `Engine` must maintain a _usage counter_ during variable assignment / creation operations.
+This adds a bit of complexity, but it should not affect performance, as in most cases, the `Engine` will only have to check a flag and only in very rare cases do extra work.
+Even an `IDisposable` check needs to be done only once.
+Moreover, we can even make a cache for all known types using our `TypeGen.ps1` utility.
+
+Another feature is that native objects can turn around in `PSObject` and the `Engine` should understand this too.
+
+### Annotating types / objects
+
+There are four options for annotating:
+
+- make a cache for all known types using our `TypeGen.ps1` utility
+- use ETS to enhance some types with this feature with default options
+
+    For example, if we define a TempFile type then users will not have to worry about deleting temporary files created by New-TemporaryFile cmdlet at all
+- embed this in the cmdlets so that they issue an already annotated object into the pipeline
+
+    For example, the same New-TemporaryFile or Open-Database cmdlet could do the annotation internally so that users have no need do this explicitly
+- explicitly annotate like `$tempFile.MakeDisposable(DisposableKind.AtRunspaceClose, { script to remove the temp file })`
+
+Note that there are many APIs that use disposable classes and now developers will have the opportunity to freely write objects of these types in the pipeline without fear of resource leakage.
+
+### Kinds of annotation
+
+Mentioned `DisposableKind` is based on context / scope and could be:
+
+- `Unknown`, it is default for new object. The `Engine` should annoutate the object. The default allow the `Engine` to process old code.
+- `Auto`, the `Engine` annoutates based on current context/scope
+- `AtGlobal`, dispose at default runspace close
+- `AtModuleUnload`
+- `AtRunspaceClose`
+- `AtCmdletClose/AtPipelineClose`
+- `AtScriptBlockClose`, dispose at current script block scope close
+
+Thus, a developer is always able to set a desired behavior.
+
+Moreover, we can always call MakeDisposable() in the script to adjust the behavior if the `Engine` cannot do this automatically.
+In this case, we will also need a method to force the mandatory code to execute like `$tempFile.PSDispose()`.
+It may also be required if we want to free an expensive resource before an end of the script block.
+
+### Implementtation details
+
+The main difficulty is that this is a generalized solution and users will expect it to work everywhere and as expected.
+Thus, it is possible the implementation will be adjusted based on feedback for a long time.
```

Comment:
This statement is concerning to me. This means that we will potentially have several versions of PowerShell with differing behaviour here, with no changes to the script they're being asked to run, correct?

That sounds like a very long series of breaking changes and a general unpredictability of how a given script would operate across PowerShell versions.

Reply:
Experimental feature will dispel your concern.

Reply:
Only if we can be absolutely certain the feature is modular enough to be hidden in a feature flag.